mlkem/src/native/riscv64/src/rv64v_poly.c

```diff
@@ -0,0 +1,750 @@
+  MLK_RVV_BFLY_FX(v4, vc, vt, zeta[0x01], vl);
+  MLK_RVV_BFLY_FX(v5, vd, vt, zeta[0x01], vl);
+  MLK_RVV_BFLY_FX(v6, ve, vt, zeta[0x01], vl);
+  MLK_RVV_BFLY_FX(v7, vf, vt, zeta[0x01], vl);
+
+  MLK_RVV_BFLY_FX(v0, v4, vt, zeta[0x10], vl);
+  MLK_RVV_BFLY_FX(v1, v5, vt, zeta[0x10], vl);
+  MLK_RVV_BFLY_FX(v2, v6, vt, zeta[0x10], vl);
+  MLK_RVV_BFLY_FX(v3, v7, vt, zeta[0x10], vl);
+  MLK_RVV_BFLY_FX(v8, vc, vt, zeta[0x11], vl);
+  MLK_RVV_BFLY_FX(v9, vd, vt, zeta[0x11], vl);
+  MLK_RVV_BFLY_FX(va, ve, vt, zeta[0x11], vl);
+  MLK_RVV_BFLY_FX(vb, vf, vt, zeta[0x11], vl);
+
+  MLK_RVV_BFLY_FX(v0, v2, vt, zeta[0x20], vl);
+  MLK_RVV_BFLY_FX(v1, v3, vt, zeta[0x20], vl);
+  MLK_RVV_BFLY_FX(v4, v6, vt, zeta[0x21], vl);
+  MLK_RVV_BFLY_FX(v5, v7, vt, zeta[0x21], vl);
+  MLK_RVV_BFLY_FX(v8, va, vt, zeta[0x30], vl);
+  MLK_RVV_BFLY_FX(v9, vb, vt, zeta[0x30], vl);
+  MLK_RVV_BFLY_FX(vc, ve, vt, zeta[0x31], vl);
+  MLK_RVV_BFLY_FX(vd, vf, vt, zeta[0x31], vl);
+
+  MLK_RVV_BFLY_FX(v0, v1, vt, zeta[0x40], vl);
+  MLK_RVV_BFLY_FX(v2, v3, vt, zeta[0x41], vl);
+  MLK_RVV_BFLY_FX(v4, v5, vt, zeta[0x50], vl);
+  MLK_RVV_BFLY_FX(v6, v7, vt, zeta[0x51], vl);
+  MLK_RVV_BFLY_FX(v8, v9, vt, zeta[0x60], vl);
+  MLK_RVV_BFLY_FX(va, vb, vt, zeta[0x61], vl);
+  MLK_RVV_BFLY_FX(vc, vd, vt, zeta[0x70], vl);
+  MLK_RVV_BFLY_FX(ve, vf, vt, zeta[0x71], vl);
+
+  __riscv_vse16_v_i16m2(
+      &r[0x00], mlk_rv64v_ntt2(__riscv_vcreate_v_i16m1_i16m2(v0, v1), z0), vl2);
+  __riscv_vse16_v_i16m2(
+      &r[0x20], mlk_rv64v_ntt2(__riscv_vcreate_v_i16m1_i16m2(v2, v3), z2), vl2);
+  __riscv_vse16_v_i16m2(
+      &r[0x40], mlk_rv64v_ntt2(__riscv_vcreate_v_i16m1_i16m2(v4, v5), z4), vl2);
+  __riscv_vse16_v_i16m2(
+      &r[0x60], mlk_rv64v_ntt2(__riscv_vcreate_v_i16m1_i16m2(v6, v7), z6), vl2);
+  __riscv_vse16_v_i16m2(
+      &r[0x80], mlk_rv64v_ntt2(__riscv_vcreate_v_i16m1_i16m2(v8, v9), z8), vl2);
+  __riscv_vse16_v_i16m2(
+      &r[0xa0], mlk_rv64v_ntt2(__riscv_vcreate_v_i16m1_i16m2(va, vb), za), vl2);
+  __riscv_vse16_v_i16m2(
+      &r[0xc0], mlk_rv64v_ntt2(__riscv_vcreate_v_i16m1_i16m2(vc, vd), zc), vl2);
+  __riscv_vse16_v_i16m2(
+      &r[0xe0], mlk_rv64v_ntt2(__riscv_vcreate_v_i16m1_i16m2(ve, vf), ze), vl2);
+}
+
+#undef MLK_RVV_BFLY_FX
+#undef MLK_RVV_BFLY_FV
+
+/*************************************************
+ * Name:        poly_invntt_tomont
+ *
+ * Description: Computes inverse of negacyclic number-theoretic transform (NTT)
+ *              of a polynomial in place;
+ *              inputs assumed to be in bitreversed order,
+ *              output in normal order
+ *
+ * Arguments:   - uint16_t *r: pointer to in/output polynomial
+ **************************************************/
+
+/*  reverse butterfly operation */
+
+#define MLK_RVV_BFLY_RX(u0, u1, ut, uc, vl) \
+  {                                         \
+    ut = __riscv_vsub_vv_i16m1(u0, u1, vl); \
+    u0 = __riscv_vadd_vv_i16m1(u0, u1, vl); \
+    u0 = fq_csub(u0, vl);                   \
```

Comment:
@mkannwischer @mjosaarinen Overall, the implementations seems to work with signed rather than unsigned values, hence I would expect every normalization operation to be symmetric -- but here we are only reducing values > Q, while keeping values < -Q as they are. That would suggest that with bad luck, one could get negative arithmetic overflow.

Can you clarify?